### ISSUE TITLE
Clarification on What is needed in Customized Objective Function

### ISSUE THREAD

**Issue by tqchen** [dmlc/xgboost#1825]
I always hopped that I could write a tutorial on how exactly customized objective functions works. But did not get time to do it. This issue is a brief summary of what do we need exactly in customized objective function. Hopefully in time, we could, or some volunteers would refine based on this and we can bring it into xgboost's document like the introduction to boosted trees

## Old Story: Second Order Approximation
To get a customized objective function, we want to define a function that returns gradient, and second order gradient statistics for each instance.

In our old tutorial, we explained that it is a second order Taylor expansion to the original objective function, and we iteratively optimizes the approximated function at each boosting iterations

However, there are cases where it falls apart:
- There are cases where the objective is not second order differentiable
- There are cases where the objective is second order differentiable, but leads to unconventional behavior.

## The True Convergence Condition: Upper Bound

The true condition behind gradient boosting to converge is to derive an upper bound of original function. This works exactly like [Majorization-Minimization](https://en.wikipedia.org/wiki/MM_algorithm) or EM style algorithm. Where we tries to find a quadratic upper bound of the target objective function, and minimize the upper bound. See Thm 4.3 for [this paper](http://homes.cs.washington.edu/~tqchen/data/pdf/GBCRF-AISTATS15.pdf) under a variant of gradient boosting.

When we return g, h pair in the custom function, we actually want to find g and h such that.
```
obj <= const + g dy + 0.5 * h dy^2 
```

The reason why second-order expansion works, is because it approximately tightly upper bound the objective function(given that the delta step is small). 

This also explains why it fails sometimes, especially in some-cases the second order gradient is zero.
This is because the update delta step is so large, such that the inequality is no longer true. In these cases, we also want to set ```max_delta_step```, so the inquality only have to hold within that region.

**Comment by tqchen** [comment]
Again, it would be nice to illustrate this with plots, and curves like the explanations for EM algorithm, and refine the explanation. If you are interested in helping out. Please follow up this issue and post your refinements(and figures) in here.

**Comment by pommedeterresautee** [comment]
Hi @tqchen , I would be interested in writing this tuto. I totally see how nice it is to present boosting as a kind of MM algo. I have a technical question (and probably other will follow):
MM (minimization) requires that the surrogate function is >= the original function to minimize. 
How do you guarantee that the Taylor expansion of 2nd degree is >= to the original loss?

**Comment by tqchen** [comment]
yes, it is approximately bigger equal than the original loss

**Comment by tqchen** [comment]
https://homes.cs.washington.edu/~tqchen/data/pdf/GBCRF-AISTATS15.pdf contains some derivations on bounds related to softmax
